New routes and various fixes.

dormando opened the pull request:
- Makes failover zone aware (see commit/test)
- Specifying "internal" as child uses the process-local cache.
- Allow `backend_options` in `settings{}` global overrides
- Adds `route_ratelim` (WIP)
- Adds `route_allasync` to dispatch requests but immediately return (same results as nullroute)

Requires `next` branch of memcached.

TODO:

- need to solve for specifying "a set but without the local zone" and "a set but just the local zone" - maybe just as an option to the routes.
this lets you specify a failover route with local zone, then an allfastest route as the failover child. which then runs all possible failover pools in parallel.

- finish ratelim
hook it up with a nullroute res and do some tests. should be possible to mix with failover to ratelimit access to new thing etc.
should also add an option to invert: return fail until ratelimit has been hit. this would allow you to do a scenario like:
- failover between A and B (then maybe A again). A is remote pool. B is internal pool.
- if rate is below N send gets to A 
- if rate is above N send gets to internal process for extra speed (but potential staleness/etc).
- there other potential uses but just throwing out an example.

dormando commented:
@sergmour - tagging you here to follow if you want.

I'm adding a couple more options then merging this and cutting a new mc tarball for it.

The `route_zfailover` will be deprecated. You should be able to do what it did by mixing failover and allfastest. (once I figure out one or two more things with it).

sergmour commented:
@dormando : Thanks! This makes me happy as-is. Below are some considerations from the top of my head:

- `route_internal` uses the same Memcached-proxy process as a backend? This is a good use case for a short-lived L0 cache.
- Have you considered having more than one local zone? Our case is that we have 3 pools each holding a dataset per region (3 near pools) and several far pools (usually 3+ regions each also having 3 pools). Read requests are localized in the same region. Mutations go to all. This is why we create a separate config-files for each pool currently using Mcrouter. This way we always know the proper order of children and timeouts. Local_zone covers this to a certain extent but it might not be flexible enough. This is just a consideration.
- Have you considered a function to determine local zone dynamically? I don't think it is easy to do in a generic way, just a consideration.
- `route_ratelim` is supposed to be used with other nested routes, right? I wonder where this delineation between separate route handles and options should happen. It is probably easier to use a combination of routes - less repeatable code versus options.
- Shuffling... Have you considered a route_random implementation? I wonder if this would solve the shuffling predicament by putting local_zone first and random_route encapsulating the far zones. This would also help randomize multiple local zones if we decide on multiple local zones.
- Logging is great. I haven't looked deeply but I think Memcached-proxy lacks latency metrics. At least we can emit a metric for anything beyond the `deadline` threshold.

dormando commented:
@sergmour yes. any proxy can also be storage, and any storage can be a proxy. it's supposed to be the big fancy feature of the proxy but it's taken a long time to get it exposed in an easy to use interface. In routelib it's a magic pool called "internal" that you can pass to any route handler. So you can run failover from internal to remote, or allfastest with internal + remote or etc.

Another example is taking an existing pool and enabling proxy mode such that the pool copies sets to another AZ/DC (and possibly fetches misses through it) for high speed situations that don't want to use an extra router hop.

- re local zones. it's three full copies of cache in each region, so 12+ copies in total? or is this "3 servers as one pool in one region with the cache copied between all of them"? I think it should be easy to build this as a concept but you might have to do custom route handles.

- re dynamic: define dynamic? It's possible to pull it in via an environment variable, from a file, or etc. I was planning on adding some convenience functions for that in this PR (`local_zone_from_env("ENVVAR")`, etc). If that's what you were thinking.

- route_ratelim: You can nest it above/below other routes or use it to rate limit access to a single pool. I'm sure some other routes may or may not use rate limits internally (like slow warmup/etc) whenever we get to that. It's using an internal API that's been around a long time.

- random: no reason not to but I don't have a good vision for how it'd be used offhand.

- you can set the logs to emit 1 out of N lines + every line over deadline and get an average + worst case. there're a ton of other metrics and metric generators in the base API... figuring out how to expose it is going to take a while.

I'm trying to prioritize getting the lua API fully documented so it might be more obvious what kind of routes you can write, anyway.

sergmour commented:
- Re local zone: Yes, 12 full copies of cache spread between 3 regions. The latency between regions is high enough. That's why we localize reads within one region with 3x redundancy (sometimes 2x).
- Re dynamic, yes, this is exactly what I meant. In this case the same config-file can be used in different regions.
- Re random, we use Mcrouter random route to evenly distribute writes to Mcrouters in external cluster (for sync/migration and other purposes).
- Thanks for the hint about metrics. Yes, that's what we lack now before we fully switch to memcached-proxy.
- Another thing that is missing versus Mcrouter, big-value-split feature. It is helpful but it needs some effort to implement properly in memcached-proxy.

dormando commented:
@sergmour any chance you could go into detail on how/why you use bigvaluesplit? It didn't seem like anyone outside of FB actually used it. Possibly in a separate issue or discussion?

I have some API pieces that could be used for that but it would still require buffering the entire value on send/receive. Since OSS memcached can raise the item limiit with `-I` that's been good enough for the use cases so far.

Also separately, specifically what metrics were you relying on? (what command/etc to get them out of mcrouer and what did they look like?) - I'll see if there's already an equivalent hiding in there.

sergmour commented:
Will create an issue for big-value-split by the end of the day and will also give you the examples on metrics.

dormando commented:
Think I'll elect to not solve the zone issue in this particular PR. I need to cut a tarball release and want to pull in these routelib updates. Will keep hacking at it presently and can cut more releases at any time.

Think it might be fair to add a few more options to failover without it being nonsensical:
- `wait = fractional_seconds` -- failover if took more than N seconds
- `parallel = true` -- basically switch to what zfailover was, but without fastgood. this "solves" the zone issue in a way.

Think before merge I'll try to go for:

- `route_random` -- should be super trivial
- `wait` could also be implemented as `route_wait` for composition purposes.

Maybe:
- backoff options for failover. This is implemented down deep for when backends _actually_ fail vs are just slow so it's not critical. If doing it high level, using wait limits instead of hard-resets on backends, we can do things like random-chance fail to do some load balancing.

dormando commented:
think I'm done for this release... I added `wait` to allfastest too and will do `parallel` on failover another time.

pretty big bump in capability overall this week so I'm happy with it. Just need to make sure we get monthly improvements from now on...

dormando commented:
@sergmour this is merged now. think I've addressed most of your concerns, except:
- think we still need a separate issue on metrics you're expecting.
- re: "multiple local zones" - I've punted on this for now. Worth noting `local_zone` is purely a routelib concept (ie; pure lua scripting) so we have flexibility in solving this.

sergmour commented:
@dormando : Will create an issue on metrics tomorrow. Re multiple local zones, yes, agree, all can be done in routelib. Another thing I wondered about is a concept of plugged-in user code upon receiving a request and upon sending the response back. It's probably an overkill for now but just a concept for future.

dormando commented:
@sergmour can you rephrase or give an example? I think this is just custom route handlers.

sergmour commented:
@dormando : Right, makes sense. Putting a custom handle on top to do what you want and the predefined route handles underneath. Yes, pretty flexible.

dormando commented:
@sergmour yeah that's the right idea. there's a customroute example in the repo. you put your own file in there (which can further load other files via lua `dofile` commands). If you issue a config reload it'll load the new code as well, so you can dynamically change the behavior of the proxy (without affecting in-flight requests).

You'll lose speed with each extra route chained but I'm hoping with 1.6.35 we're at an acceptable speed.
